core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -492,6 +455,64 @@ private[kafka] class Processor(val id: Int,
     }
   }
 
+  /* `protected` for test usage */
+  protected[network] def sendResponse(response: RequestChannel.Response) {
+    trace(s"Socket server received response to send, registering for write and sending data: $response")
+    selector.send(response.responseSend)
```

Comment:
It is possible that the client already closed the socket when the response is ready. In this case, selector.send() will hit an IllegalStateException if the channel is already removed from Selector, but we still want to update the metrics.